### PR TITLE
Book category

### PR DESCRIPTION
**Pull request by dmytrocherepov**
added book category , service , mapper , liquibase

### CODE REVIEW

File: src/main/java/com/example/bookstoreapp/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category management ", description = "Endpoints for managing categories")
+@RestController
+@RequestMapping("/categories")
+@RequiredArgsConstructor
+public class CategoryController {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/controller/CategoryController.java]:
```suggestion
```

File: src/main/java/com/example/bookstoreapp/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/controller/CategoryController.java]:
```suggestion
    @PreAuthorize("hasRole('ADMIN')")
```
you can simplify  it, apply to other endpoints

File: src/main/java/com/example/bookstoreapp/dto/book/CreateBookRequestDto.java [hunk 3]
```diff
@@ -21,7 +22,9 @@ public record CreateBookRequestDto(
         @NotNull
         BigDecimal price,
         String description,
-        String coverImage
+        String coverImage,
+        @NotNull
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/dto/book/CreateBookRequestDto.java]:
```suggestion
        @NotEmpty
```

File: src/main/java/com/example/bookstoreapp/exception/CategoryException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.example.bookstoreapp.exception;
+
+public class CategoryException extends RuntimeException {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/exception/CategoryException.java]: Don’t  create such exception just use EntityNotFoundException

File: src/main/java/com/example/bookstoreapp/mapper/CategoryMapper.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.example.bookstoreapp.config.MapperConfig;
+import com.example.bookstoreapp.dto.category.CategoryRequestDto;
+import com.example.bookstoreapp.dto.category.CategoryResponseDto;
+import com.example.bookstoreapp.model.Category;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface CategoryMapper {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/mapper/CategoryMapper.java]:
```suggestion
```

File: src/main/java/com/example/bookstoreapp/model/Book.java [hunk 6]
```diff
@@ -30,6 +35,14 @@ public class Book {
     private BigDecimal price;
     private String description;
     private String coverImage;
+
+    @ManyToMany
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/model/Book.java:41]: if you use data better add 
```
@ToString.Exclude
@EqualsAndHashCode.Exclude
```

File: src/main/java/com/example/bookstoreapp/repository/category/CategoryRepository.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.example.bookstoreapp.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    Boolean existsByName(String name);
+
+    void deleteById(Long id);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/repository/category/CategoryRepository.java]: is it mandatory to record these methods in the repository?

File: src/main/java/com/example/bookstoreapp/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -64,4 +73,18 @@ public List<BookDto> search(BookSearchParameters params, Pageable pageable) {
...
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> findBooksByCategory(Pageable pageable, Long id) {
+        return bookRepository.findBooksByCategoriesId(id).stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .toList();
+    }
+
+    private void setCategories(CreateBookRequestDto requestDto, Book book) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/BookServiceImpl.java:86]: you should also add it to update method

File: src/main/java/com/example/bookstoreapp/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import com.example.bookstoreapp.service.CategoryService;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/CategoryServiceImpl.java:18]:
```suggestion
```

File: src/main/java/com/example/bookstoreapp/service/impl/CategoryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+                .map(categoryMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException("No such category"));
+    }
+
+    @Override
+    public CategoryResponseDto save(CategoryRequestDto categoryDto) {
+        Category entity = categoryMapper.toEntity(categoryDto);
+        if (categoryRepository.existsByName(categoryDto.name())) {
+            throw new CategoryException("Category already exists");
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/CategoryServiceImpl.java]: Interesting 🤔  but don’t  complicate your solution)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
There are merge conflicts. Pls resolve them first :)

**Comment by dmytrocherepov** [comment]
solved conflicts

**Pull request merged**